From 27ff29e27d4f77a4c00b60f06ffaeacfdb4acea7 Mon Sep 17 00:00:00 2001
From: Andreas Beckmann <anbe@debian.org>
Date: Tue, 3 Jan 2023 21:20:42 +0100
Subject: [PATCH] support acpi_op_remove callback returning void

returns void instead of int since 6c0eb5ba3500f (in v6.2-rc1)
---
 conftest.sh      | 18 ++++++++++++++++++
 nvidia/nv-acpi.c |  6 ++++++
 2 files changed, 24 insertions(+)

diff --git a/conftest.sh b/conftest.sh
index 88c7eb1..b9574f9 100755
--- a/conftest.sh
+++ b/conftest.sh
@@ -863,6 +863,24 @@ compile_test() {
                 return
             fi
 
+            echo "$CONFTEST_PREAMBLE
+            #include <linux/acpi.h>
+
+            acpi_op_remove conftest_op_remove_routine;
+
+            void conftest_acpi_device_ops_remove(struct acpi_device *device) {
+                return conftest_op_remove_routine(device);
+            }" > conftest$$.c
+
+            $CC $CFLAGS -c conftest$$.c > /dev/null 2>&1
+            rm -f conftest$$.c
+
+            if [ -f conftest$$.o ]; then
+                rm -f conftest$$.o
+                echo "#define NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT -1" | append_conftest "types"
+                return
+            fi
+
             CODE="
             #include <linux/acpi.h>
 
diff --git a/nvidia/nv-acpi.c b/nvidia/nv-acpi.c
index da9fbc9..06709e6 100644
--- a/nvidia/nv-acpi.c
+++ b/nvidia/nv-acpi.c
@@ -25,6 +25,8 @@ static int         nv_acpi_add             (struct acpi_device *);
 
 #if !defined(NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT) || (NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT == 2)
 static int         nv_acpi_remove_two_args(struct acpi_device *device, int type);
+#elif (NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT == -1)
+static void        nv_acpi_remove_one_arg(struct acpi_device *device);
 #else
 static int         nv_acpi_remove_one_arg(struct acpi_device *device);
 #endif
@@ -333,6 +335,8 @@ static int nv_acpi_add(struct acpi_device *device)
 
 #if !defined(NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT) || (NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT == 2)
 static int nv_acpi_remove_two_args(struct acpi_device *device, int type)
+#elif (NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT == -1)
+static void nv_acpi_remove_one_arg(struct acpi_device *device)
 #else
 static int nv_acpi_remove_one_arg(struct acpi_device *device)
 #endif
@@ -385,7 +389,9 @@ static int nv_acpi_remove_one_arg(struct acpi_device *device)
         device->driver_data = NULL;
     }
 
+#if !defined(NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT) || (NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT != -1)
     return status;
+#endif
 }
 
 static void nv_acpi_event(acpi_handle handle, u32 event_type, void *data)
-- 
2.20.1

